smithed_libraries/plugins/versioning/models.py

```diff
@@ -0,0 +1,114 @@
+
+from beet import Context, ListOption, PathSpecOption
+from beet.contrib.rename_files import RenderRenameOption, TextRenameOption
+from pydantic import BaseModel, Field, validator
+
+from .types import JsonDict, JsonType
+
+
+class ContextualModel(BaseModel):
+    ctx: ClassVar[Context]
+
+
+class Version(ContextualModel):
+    """Provides methods and fields for the version for ease of use"""
+
+    __root__: dict[str, int]
+
+    @classmethod
+    def from_parts(cls, names: list[str]):
+        numbers = [int(num) for num in cls.ctx.project_version.split(".")]
+        combined = dict(zip(names, numbers))
+
+        if len(combined) != len(numbers) or len(combined) != len(names):
+            raise ValueError(
+                f"The schema: {names!r} does not match the version {cls.ctx.project_version!r}"
+            )
+
+        return cls(__root__=combined)
+
+    @validator("__root__")
+    def ensure_version_parts(cls, value: dict[str, int]):
+        for name, number in value.items():
+            if number < 0:
+                raise ValueError(f"{name} had an invalid version value: {number}")
+
+        return value
+
+    def named_parts(self):
+        return list(self.__root__.items())
+
+    def __str__(self):
+        return ".".join(str(value) for value in self.__root__.values())
+
+
+class VersioningOptions(ContextualModel, extra="forbid"):
+    """The `versioning` config of `beet.yaml`"""
+
+    class LanternLoadOptions(ContextualModel):
+        step: Literal["pre_load", "load", "post_load"] = "load"
+        tag_path: str = "load"
+        function_path: str = "impl/load"
+
+    class ApiOptions(ContextualModel):
+        match: PathSpecOption = "{{ project_id }}:v{{ version }}/*"  # type: ignore
+        implementation_prefix: str = "{{ project_id }}:v{{ version }}/"
+        version_check_path: str = "v{{ version }}/calls"
+        tag_path: str = ""
+
+    scoreholder: str = "#{{ project_id }}"
+    schema_: list[str] = Field(["major", "minor", "patch"], alias="schema")
+    scheduled_paths: ListOption[str] = ListOption(__root__=["impl/tick"])
+    version: Version = None  # type: ignore
+    refactor: TextRenameOption | RenderRenameOption = {
+        "match": "{{ project_id }}:*",
+        "find": "{{ project_id }}:impl/",
+        "replace": "{{ project_id }}:impl/v{{ version }}/",
+    }  # type: ignore
+    lantern_load: LanternLoadOptions = LanternLoadOptions()
+    api: ApiOptions = ApiOptions()
+
+    @property
+    def namespace(self):
+        return self.ctx.project_id
+
+    @classmethod
+    def render(cls, value: str, values: dict[str, Any]):
+        return cls.ctx.template.render_string(value, **values)
+
+    @validator("version", pre=True, always=True)
+    def init_version(cls, value: Any, values: dict[str, Any]):
+        match value:
+            case None:
+                return Version.from_parts(values["schema_"])
+            case _:
+                return value
```

Comment:
Unless you're using the match statement deliberately this would probably be easier to read as `return value or Version.from_parts(values["schema_"])`.